File: mix/python/pitch_map.py
"""Continuous evidence from bass inference, independent of instrument and note layout.

No pitch is snapped or corrected here. A rejected estimate is retained as evidence,
not presented as a voiced fundamental. Periodicity is not a probability of accuracy.
"""

import hashlib
import json
from pathlib import Path

import numpy as np

FORMAT = 1
FILE = "pitch-map.json"
SILENCE_DB = -60.0


def file_hash(file: Path) -> str:
    with file.open("rb") as stream:
        return hashlib.file_digest(stream, "sha256").hexdigest()


def frame_step(sample_rate: int, hop_length: int) -> float:
    """torchcrepe preprocess truncates the hop after resampling to 16 kHz."""
    return int(hop_length * 16000 / sample_rate) / 16000


def make_map(mono, sample_rate, hop_length, hz, periodicity, smoothed, threshold):
    step = frame_step(sample_rate, hop_length)
    if step <= 0 or len(hz) != len(periodicity) or len(hz) != len(smoothed):
        raise ValueError("invalid pitch frame clock or array lengths")
    duration = len(mono) / sample_rate
    # pad=True centers the first 64 ms model window on sample zero. Do not
    # keep a padded endpoint at or beyond the source's end.
    times = np.arange(len(hz)) * step
    keep = times < duration
    times, hz, periodicity, smoothed = times[keep], hz[keep], periodicity[keep], smoothed[keep]
    centres = np.rint(times * sample_rate).astype(int)
    half = round(0.032 * sample_rate)
    lo, hi = np.maximum(0, centres - half), np.minimum(len(mono), centres + half)
    energy = np.concatenate(([0.0], np.cumsum(np.square(mono, dtype=np.float64))))
    rms = np.sqrt(np.maximum(0, energy[hi] - energy[lo]) / np.maximum(1, hi - lo))
    db = 20 * np.log10(np.maximum(rms, 1e-12))
    finite = np.isfinite(hz) & (hz > 0) & np.isfinite(periodicity) & np.isfinite(smoothed)
    state = np.full(len(hz), "unvoiced", dtype=object)
    state[finite & (smoothed >= threshold)] = "voiced"
    state[db < SILENCE_DB] = "silent"
    state[~finite] = "invalid"

    def values(array):
        return [float(v) if np.isfinite(v) else None for v in array]

    return {
        "openflow": "mix-pitch-map", "version": FORMAT,
        "seconds": duration, "sampleRate": sample_rate,
        "start": 0, "step": step, "windowSeconds": 0.064,
        "policy": {"periodicityFloor": threshold, "silenceDb": SILENCE_DB},
        "hz": values(hz), "periodicity": values(periodicity),
        "smoothedPeriodicity": values(smoothed), "rmsDb": values(db),
        "state": state.tolist(),
    }


def save_map(out, source, source_hash, engine, **frames):
    held = make_map(**frames)
    held.update(source={"hash": source_hash or file_hash(source), "bytes": source.stat().st_size}, engine=engine)
    body = json.dumps(held, separators=(",", ":"), allow_nan=False).encode()
    out.mkdir(parents=True, exist_ok=True)
    (out / FILE).write_bytes(body)
    return {"file": FILE, "version": FORMAT, "frames": len(held["hz"]),
            "sha256": hashlib.sha256(body).hexdigest()}

File: mix/src/pitchMap.ts
import type { TranscribedNote } from './tab.ts';

export interface PitchEvidence { map: PitchMap; notes: TranscribedNote[]; transpose: number }

/** Raw model evidence. Instrument tuning and manual note corrections do not alter it. */
export interface PitchMap {
  openflow: 'mix-pitch-map';
  version: 1;
  source: { hash: string; bytes: number };
  engine: { name: string; version: string; model: string; fmin: number; fmax: number; hopMs: number; decoder: string; batchSize: number; seed: number };
  seconds: number;
  sampleRate: number;
  start: number;
  step: number;
  windowSeconds: number;
  policy: { periodicityFloor: number; silenceDb: number };
  hz: (number | null)[];
  periodicity: (number | null)[];
  smoothedPeriodicity: (number | null)[];
  rmsDb: (number | null)[];
  state: ('voiced' | 'unvoiced' | 'silent' | 'invalid')[];
}

export const midiPitch = (hz: number): number => 69 + 12 * Math.log2(hz / 440);
const NAMES = ['C', 'C♯', 'D', 'E♭', 'E', 'F', 'F♯', 'G', 'A♭', 'A', 'B♭', 'B'];
export const pitchName = (pitch: number): string => `${NAMES[((Math.round(pitch) % 12) + 12) % 12]}${Math.floor(Math.round(pitch) / 12) - 1}`;

/** Reject malformed evidence rather than draw plausible frames on a wrong clock. */
export function validPitchMap(value: unknown): value is PitchMap {
  if (!value || typeof value !== 'object') return false;
  const p = value as PitchMap;
  if (p.openflow !== 'mix-pitch-map' || p.version !== 1 || p.start !== 0 ||
    !Number.isFinite(p.seconds) || p.seconds <= 0 || !Number.isFinite(p.step) || p.step <= 0 ||
    !Number.isFinite(p.sampleRate) || p.sampleRate <= 0 || !Number.isFinite(p.windowSeconds) || p.windowSeconds <= 0 ||
    typeof p.source?.hash !== 'string' || !Number.isFinite(p.source.bytes) ||
    typeof p.engine?.name !== 'string' || !Number.isFinite(p.engine.fmin) || !Number.isFinite(p.engine.fmax) ||
    !Number.isFinite(p.policy?.periodicityFloor) || !Number.isFinite(p.policy.silenceDb) ||
    !Array.isArray(p.hz) || p.hz.length === 0) return false;
  const n = p.hz.length;
  if ((n - 1) * p.step >= p.seconds || n * p.step < p.seconds - p.step) return false;
  const numbers = [p.hz, p.periodicity, p.smoothedPeriodicity, p.rmsDb];
  if (!numbers.every(a => Array.isArray(a) && a.length === n && a.every(v => v === null || Number.isFinite(v)))) return false;
  return Array.isArray(p.state) && p.state.length === n && p.state.every((s, i) =>
    ['voiced', 'unvoiced', 'silent', 'invalid'].includes(s) &&
    (s !== 'voiced' || (p.hz[i] !== null && p.hz[i]! > 0 && p.smoothedPeriodicity[i] !== null)));
}
